src/storage/src/monitor/traced_store.rs

```diff
@@ -0,0 +1,244 @@
+            trace_result!(GET, span, res);
+            res
+        }
+    }
+
+    fn iter(
+        &self,
+        key_range: (Bound<Vec<u8>>, Bound<Vec<u8>>),
+        epoch: u64,
+        read_options: ReadOptions,
+    ) -> Self::IterFuture<'_> {
+        async move {
+            let span: TraceSpan = trace!(ITER, key_range, epoch, read_options, self.storage_type);
+            let iter = self
+                .traced_iter(
+                    self.inner.iter(key_range, epoch, read_options),
+                    span.id(),
+                    span,
+                )
+                .await;
+            iter
+        }
+    }
+}
+
+impl<S: StateStoreWrite> StateStoreWrite for TracedStateStore<S> {
+    define_state_store_write_associated_type!();
+
+    fn ingest_batch(
+        &self,
+        kv_pairs: Vec<(Bytes, StorageValue)>,
+        delete_ranges: Vec<(Bytes, Bytes)>,
+        write_options: WriteOptions,
+    ) -> Self::IngestBatchFuture<'_> {
+        async move {
+            // Don't trace empty ingest to save disk space
+            if kv_pairs.is_empty() && delete_ranges.is_empty() {
+                return Ok(0);
+            }
+
+            let span: TraceSpan = trace!(
+                INGEST,
+                kv_pairs,
+                delete_ranges,
+                write_options,
+                self.storage_type
+            );
+            let res: StorageResult<usize> = self
+                .inner
+                .ingest_batch(kv_pairs, delete_ranges, write_options)
+                .await;
+            trace_result!(INGEST, span, res);
+            res
+        }
+    }
+}
+
+impl TracedStateStore<HummockStorage> {
+    pub fn sstable_store(&self) -> SstableStoreRef {
+        self.inner.sstable_store()
+    }
+
+    pub fn sstable_id_manager(&self) -> &SstableIdManagerRef {
+        self.inner.sstable_id_manager()
+    }
+}
+
+pub struct TracedStateStoreIter<I> {
+    inner: I,
+    record_id: RecordId,
+    span: TraceSpan,
+}
+
+impl<I> TracedStateStoreIter<I> {
+    fn new(inner: I, record_id: RecordId, span: TraceSpan) -> Self {
+        span.send_result(OperationResult::Iter(TraceResult::Ok(())));
+        TracedStateStoreIter {
+            inner,
+            record_id,
+            span,
+        }
+    }
+}
+
+impl<I> StateStoreIter for TracedStateStoreIter<I>
+where
+    I: StateStoreIter<Item = (FullKey<Vec<u8>>, Bytes)>,
+{
+    type Item = (FullKey<Vec<u8>>, Bytes);
+
+    type NextFuture<'a> = impl NextFutureTrait<'a, Self::Item>;
+
+    fn next(&mut self) -> Self::NextFuture<'_> {
+        async move {
+            if should_use_trace() {
+                self.span.send(Operation::IterNext(self.record_id));
+            }
+            let kv_pair: _ = self.inner.next().await?;
+            trace_result!(ITER_NEXT, self.span, kv_pair);
+            Ok(kv_pair)
+        }
+    }
+}
+
+pub fn get_concurrent_id() -> ConcurrentId {
+    #[cfg(all(not(madsim), hm_trace))]
```

Comment:
We may not allow doing hummock trace when `madsim` is enabled. Therefore, we don't have to do `cfg` here, since the whole `traced_store.rs` is compiled only when hummock trace is enabled.